microcluster/app.go

```diff
@@ -96,21 +97,25 @@ func (m *MicroCluster) Start(apiEndpoints []rest.Endpoint, schemaExtensions map[
 		return fmt.Errorf("Unable to start daemon: %w", err)
 	}
 
+	stop := sync.OnceFunc(func() {
+		logger.Info("Initiate daemon stop sequence")
+		d.ShutdownDoneCh <- d.Stop()
+	})
+
 	for {
 		select {
+		case err = <-d.ShutdownDoneCh:
+			close(d.ShutdownDoneCh)
+			return err
+		case <-d.ShutdownCtx.Done():
+			stop()
 		case sig := <-sigCh:
 			logCtx := logger.AddContext(logger.Ctx{"signal": sig})
 			logCtx.Info("Received signal")
 			if d.ShutdownCtx.Err() != nil {
 				logCtx.Warn("Ignoring signal, shutdown already in progress")
-			} else {
-				go func() {
```

Comment:
I like that using sync.OnceFunc and buffered channel gets rid of the go routine though :)

Reply:
One drawback of getting rid of the go routine is that if subsequent signals are sent to the daemon, because the loop will be potentially blocked on calling `stop()` for the first time, the signals wont be handled and the `logCtx.Warn("Ignoring signal, shutdown already in progress")` part wont be reached.